README.txt
Upload Reference Widget
=======================

  The Upload Reference Widget aims to be the primary mechanism for uploading a
  file when creating a content item that uses the file. Instead of embedding
  the file into the content item, this widget uploads the file as a separate
  item and automatically populates a reference in the content item. Now Plone
  can reuse image and file assets in a more user-friendly way.


The problem
-----------

  One of the biggest issues we have seen re-using file assets in the Plone
  CMS is that once a file is uploaded as an attribute of content; you can not
  easily reuse it. Take this scenerio for instance:

    - You create a 'News Item' in '/news/foo.html' and upload an image,
      'logo.jpg', that shows a picture of a corporate logo. To address this
      image you must goto a URL like, 'http://host/news/foo.html/logo.jpg'.

    - Two weeks later you create another 'News Item' in '/news/bar.html' and
      you dont want to re-upload the corporate logo. You simply want to reuse
      the existing corporate logo. Where do you find it?

    - We could browse content and goto '/news/foo.html' and inside the
      'News Item' see the 'logo.jpg'. While this is possible. It is not
      probable that someone has created a container where images are being
      uploaded. Specifically the default Plone 'News Item' implementation
      does not work this way.


The suggested solution
----------------------

  Create a new 'UploadReferenceWidget'. What this will do is be an alternative
  widget that can work in-place of a ReferenceWidget; and should retain all
  the functionality of a Reference Field/Widget. But enable someone to upload
  a 'File'. The policy on the Field/Widget could dictate that the file gets put
  into a certain folder. Take this scenerio for instance:

    - We know that all content in 'http://host/corporate/news' is commercial
      news.

    - When user is creating a 'News Item' and uploads 'logo2.jpg' into the
      Upload Reference Widget called 'image' that we will ask the policy
      Where does this image, 'logo2.jpg' go? The policy could put it in
      /images/logos.

    - The policy would have to do everything. The Widget wouldn't help
      anything. The file would be put into the '/images/logos' folder and
      then in the 'News Item' image attribute would be the reference id to
      the '/images/logos/logo2.jpg'.


Implementation details
----------------------

  The new widget has two main features:

  - Upload a new file: create a new object and make a reference to it

  - Select an existing object: make a reference to it

  When uploading a new file, the widget checks for the mimetype and create
  an instance of the appropriate content-type: File or Image. This content
  is created in the 'startup_directory' property defined in the widget.


Example usage
-------------

  Take a look in the 'demo.py' file for a simple example usage of this widget.

  The browse and select functionality is provided by
  'archetypes.referencebrowserwidget', which means you can use any of its
  current features.

  Check the "complete reference of available
  options":http://tinyurl.com/referencebrowserwidget-options online.

  It contains the list of default property values and explains its meanings.




CHANGES.txt
Changelog
=========

0.3 (unreleased)
----------------

- Added ``z3c.autoinclude``-entrypoint to mark this as a Plone-plugin.
  This avoids the need to explicitly load the zcml-slug.
  [WouterVH]

- Fix mimetype of brain by reindexing object after creation.
  [pbauer]

- Use mimetypes to guess mimetype.
  [pbauer] 

- Add German translation.
  [pbauer]

- Fix i18n-domain for "Upload another file". 
  [pbauer]

- Changed startup directory handling - reuse getStartupDirectory 
  method from ReferenceBrowserWidget.
  [naro]


0.2.1 (2010-09-22)
------------------

- Added MANIFEST.in file.
  [WouterVH]

- Fixed i18n-domain in pot-file.
  [WouterVH]

- Add Dutch translation.
  [WouterVH]

- Added locales.
  [toutpt]

- Added French translation.
  [toutpt]

- Fixed registerType. It needs a package name since Archetypes 1.3.2.
  [toutpt]


0.2 (2009-03-22)
----------------

- Added our skin layer to all skins paths.
  [deo]

- Fixed dependency registration on both metadata.xml and configure.zcml.
  [deo]


0.1-alpha (Unreleased)
----------------------

- Log that fact that a monkey patch is being applied.
  [runyaga]

- Eggified to archetypes.uploadreferencewidget and changed to use
  archetypes.referencebrowserwidget instead of Products.ATReferenceBrowserWidget
  [swampmonkey]


archetypes/uploadreferencewidget/tests/utils.py
#
# archetypes.uploadreferencewidget
# Copyright(C), 2008, Enfold Systems, Inc. - ALL RIGHTS RESERVED
#
# This software is licensed under the Terms and Conditions
# contained within the "LICENSE.txt" file that accompanied
# this software.  Any inquiries concerning the scope or
# enforceability of the license should be addressed to:
#
# Enfold Systems, Inc.
# 4617 Montrose Blvd., Suite C215
# Houston, Texas 77006 USA
# p. +1 713.942.2377 | f. +1 832.201.8856
# www.enfoldsystems.com
# info@enfoldsystems.com
#

from ZPublisher.HTTPRequest import FileUpload


class FakeRequest:
    """Dummy request object.
    """
    def __init__(self):
        self.other = {}
        self.form = {}


class FakeFileUpload(FileUpload):
    """Dummy upload object.
    """
    __allow_access_to_unprotected_subobjects__ = 1
    headers = {}
    filename = 'dummy.gif'

    def __init__(self, filename=None, headers=None, file=None):
        self.file = file
        if filename is not None:
            self.filename = filename
        if headers is not None:
            self.headers = headers

    def seek(self, *args):
        return self.file.seek(*args)

    def tell(self, *args):
        return self.file.tell(*args)

    def read(self, *args):
        return self.file.read(*args)


archetypes/uploadreferencewidget/tests/URWTestCase.py
#
# archetypes.uploadreferencewidget
# Copyright(C), 2008, Enfold Systems, Inc. - ALL RIGHTS RESERVED
#
# This software is licensed under the Terms and Conditions
# contained within the "LICENSE.txt" file that accompanied
# this software.  Any inquiries concerning the scope or
# enforceability of the license should be addressed to:
#
# Enfold Systems, Inc.
# 4617 Montrose Blvd., Suite C215
# Houston, Texas 77006 USA
# p. +1 713.942.2377 | f. +1 832.201.8856
# www.enfoldsystems.com
# info@enfoldsystems.com
#

"""Base class for integration tests, based on ZopeTestCase and PloneTestCase.

Note that importing this module has various side-effects: it registers a set of
products with Zope, and it sets up a sandbox Plone site with the appropriate
products installed.
"""

import os
import doctest
from Testing import ZopeTestCase

# Import PloneTestCase - this registers more products with Zope as a side effect
from Products.PloneTestCase import PloneTestCase

from archetypes.uploadreferencewidget.config import PRODUCT_DIR


# Set up a Plone site, and install our custom widget product
PloneTestCase.setupPloneSite(products=['archetypes.referencebrowserwidget', 
    'archetypes.uploadreferencewidget'])

OPTIONFLAGS = (doctest.REPORT_ONLY_FIRST_FAILURE |
               doctest.ELLIPSIS |
               doctest.NORMALIZE_WHITESPACE)


class IntegrationTestCase(PloneTestCase.PloneTestCase):
    """Base class for the UploadReferenceWidget product integration tests.

    This may provide specific set-up and tear-down operations, or provide
    convenience methods.
    """

    def _setup(self):
        PloneTestCase.PloneTestCase._setup(self)
        self.IMG = open(os.path.join(PRODUCT_DIR, 'tests', 'enfold.gif'))
        self.PDF = open(os.path.join(PRODUCT_DIR, 'tests', 'enfold.pdf'))

    def createMemberarea(self, name):
        # Bypass PTC's creation of a home folder for the default user
        pass


class FunctionalTestCase(IntegrationTestCase, PloneTestCase.FunctionalTestCase):
    pass


archetypes/uploadreferencewidget/tests/test_widget.py
#
# archetypes.uploadreferencewidget
# Copyright(C), 2008, Enfold Systems, Inc. - ALL RIGHTS RESERVED
#
# This software is licensed under the Terms and Conditions
# contained within the "LICENSE.txt" file that accompanied
# this software.  Any inquiries concerning the scope or
# enforceability of the license should be addressed to:
#
# Enfold Systems, Inc.
# 4617 Montrose Blvd., Suite C215
# Houston, Texas 77006 USA
# p. +1 713.942.2377 | f. +1 832.201.8856
# www.enfoldsystems.com
# info@enfoldsystems.com
#

from Testing import ZopeTestCase
from Products.CMFPlone.utils import _createObjectByType

from StringIO import StringIO
from Products.PloneTestCase import PloneTestCase
from archetypes.uploadreferencewidget.tests import URWTestCase
from archetypes.uploadreferencewidget.widget import UploadReferenceWidget

from archetypes.uploadreferencewidget.tests.utils import FakeRequest
from archetypes.uploadreferencewidget.tests.utils import FakeFileUpload


class IntegrationTests(URWTestCase.IntegrationTestCase):

    def afterSetUp(self):
        self.doc = _createObjectByType('Document', self.portal, id='doc')
        self.image = _createObjectByType('Image', self.portal, id='image')
        self.field = self.doc.getField('relatedItems')
        self.field.isMetadata = False
        self.setRoles('Manager')
        self.request = FakeRequest()

    def uploadContentToPath(self, obj, path, destination):
        field = obj.getField('relatedItems')
        fileupload = FakeFileUpload(filename='enfold.pdf', file=self.PDF)
        field.widget = UploadReferenceWidget(startup_directory=path)
        form = {
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        self.request.form.update(form)
        obj.processForm(REQUEST=self.request)
        expected = destination.objectValues('ATFile')
        result = obj.getRefs('relatesTo')
        self.assertEqual(expected, result)

    def testSelectExistingContentProcessWidget(self):
        self.field.widget = UploadReferenceWidget()
        form = {
            'relatedItems': [self.image.UID()],
            'relatedItems_option': 'select',
        }
        expected = [self.image.UID()], {}
        result = self.field.widget.process_form(self.doc, self.field, form,
                                                validating=False)
        self.assertEqual(expected, result)

    def testValidateUploadWithoutSelectAFileInARequiredField(self):
        fileupload = FakeFileUpload(filename='', file=StringIO())
        self.field.widget = UploadReferenceWidget()
        self.field.required = True
        self.field.multiValued = False

        # First force a validation error, as we don't pass anything:
        form = {
            'relatedItems': '',
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        value, errors = self.field.widget.process_form(self.doc, self.field,
                                                       form, validating=False)
        result = self.field.validate(value, self.doc, errors)
        self.failUnless('required' in result)

        # Then pass something as the field previous value, no error this time:
        form = {
            'relatedItems': 'something',
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        value, errors = self.field.widget.process_form(self.doc, self.field,
                                                       form, validating=False)
        result = self.field.validate(value, self.doc, errors)
        self.assertEqual(result, None)

        # Finally, upload a valid file. Again, no error:
        fileupload = FakeFileUpload(filename='foo.txt', file=StringIO('foo'))
        form = {
            'relatedItems': '',
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        value, errors = self.field.widget.process_form(self.doc, self.field,
                                                       form, validating=False)
        result = self.field.validate(value, self.doc, errors)
        self.assertEqual(result, None)

    def testSelectMultipleExistingContentProcessWidget(self):
        video = _createObjectByType('File', self.portal, id='video')
        self.field.widget = UploadReferenceWidget()
        form = {
            'relatedItems': [self.image.UID(), video.UID()],
            'relatedItems_option': 'select',
        }
        expected = [self.image.UID(), video.UID()], {}
        result = self.field.widget.process_form(self.doc, self.field, form,
                                                validating=False)
        self.assertEqual(expected, result)

    def testSelectExistingContentCreateReference(self):
        self.field.widget = UploadReferenceWidget()
        form = {
            'relatedItems': [self.image.UID()],
            'relatedItems_option': 'select',
        }
        self.request.form.update(form)
        self.doc.processForm(REQUEST=self.request)
        expected = [self.image]
        result = self.doc.getRefs('relatesTo')
        self.assertEqual(expected, result)

    def testSelectMultipleExistingContentCreateReferences(self):
        video = _createObjectByType('File', self.portal, id='video')
        self.field.widget = UploadReferenceWidget()
        form = {
            'relatedItems': [self.image.UID(), video.UID()],
            'relatedItems_option': 'select',
        }
        self.request.form.update(form)
        self.doc.processForm(REQUEST=self.request)
        expected = [self.image, video]
        result = self.doc.getRefs('relatesTo')
        self.assertEqual(expected, result)

    def testUploadImageCreateObjectAndReference(self):
        fileupload = FakeFileUpload(filename='enfold.gif', file=self.IMG)
        self.field.widget = UploadReferenceWidget(startup_directory='test')
        self.portal.invokeFactory('Folder', id='test')
        form = {
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        self.request.form.update(form)
        self.doc.processForm(REQUEST=self.request)
        expected = self.portal.test.objectValues('ATImage')
        result = self.doc.getRefs('relatesTo')
        self.assertEqual(expected, result)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0].getContentType(), 'image/gif')

    def testUploadFileCreateObjectAndReference(self):
        fileupload = FakeFileUpload(filename='enfold.pdf', file=self.PDF)
        self.field.widget = UploadReferenceWidget(startup_directory='test')
        self.portal.invokeFactory('Folder', id='test')
        form = {
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        self.request.form.update(form)
        self.doc.processForm(REQUEST=self.request)
        expected = self.portal.test.objectValues('ATFile')
        result = self.doc.getRefs('relatesTo')
        self.assertEqual(expected, result)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0].getContentType(), 'application/pdf')

    def testUploadKeepCurrentReferencesForMultiValuedField(self):
        fileupload = FakeFileUpload(filename='enfold.pdf', file=self.PDF)
        self.field.widget = UploadReferenceWidget(startup_directory='test')
        self.portal.invokeFactory('Folder', id='test')
        form = {
            'relatedItems': [self.image.UID()],
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        self.request.form.update(form)
        self.doc.processForm(REQUEST=self.request)
        expected = self.portal.test.objectValues() + [self.image]
        result = self.doc.getRefs('relatesTo')
        self.assertEqual(len(expected), 2)
        self.assertEqual(expected, result)

    def testUploadKeepCurrentReferencesForMultiValuedFieldWithMultipleUploads(self):
        fileupload1 = FakeFileUpload(filename='enfold.pdf', file=self.PDF)
        fileupload2 = FakeFileUpload(filename='enfold.gif', file=self.IMG)
        self.field.widget = UploadReferenceWidget(startup_directory='test')
        self.portal.invokeFactory('Folder', id='test')
        form = {
            'relatedItems': [self.image.UID()],
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload1, fileupload2],
        }
        self.request.form.update(form)
        self.doc.processForm(REQUEST=self.request)
        expected = self.portal.test.objectValues() + [self.image]
        result = self.doc.getRefs('relatesTo')
        self.assertEqual(len(expected), 3)
        self.assertEqual(expected, result)

    def testUploadReplaceCurrentReferencesForSingleValuedField(self):
        fileupload = FakeFileUpload(filename='enfold.pdf', file=self.PDF)
        self.field.multiValued = False
        self.field.widget = UploadReferenceWidget(startup_directory='test')
        self.portal.invokeFactory('Folder', id='test')
        form = {
            'relatedItems': self.image.UID(),
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        self.request.form.update(form)
        self.doc.processForm(REQUEST=self.request)
        expected = self.portal.test.objectValues()
        result = self.doc.getRefs('relatesTo')
        self.assertEqual(expected, result)

    def testUploadKeepReferenceWhenUploadingWithoutAValueForMultiValuedField(self):
        fileupload = FakeFileUpload(filename='', file=StringIO())
        self.field.widget = UploadReferenceWidget(startup_directory='test')
        self.portal.invokeFactory('Folder', id='test')
        form = {
            'relatedItems': [self.image.UID()],
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        self.request.form.update(form)
        self.doc.processForm(REQUEST=self.request)
        self.assertEqual(self.portal.test.objectValues(), [])
        expected = [self.image]
        result = self.doc.getRefs('relatesTo')
        self.assertEqual(expected, result)

    def testUploadKeepReferenceWhenUploadingWithoutAValueForSingleValuedField(self):
        fileupload = FakeFileUpload(filename='', file=StringIO())
        self.field.multiValued = False
        self.field.widget = UploadReferenceWidget(startup_directory='test')
        self.portal.invokeFactory('Folder', id='test')
        form = {
            'relatedItems': self.image.UID(),
            'relatedItems_option': 'upload',
            'relatedItems_file': [fileupload],
        }
        self.request.form.update(form)
        self.doc.processForm(REQUEST=self.request)
        self.assertEqual(self.portal.test.objectValues(), [])
        expected = [self.image]
        result = self.doc.getRefs('relatesTo')
        self.assertEqual(expected, result)

    def testUploadCreateObjectOnCurrentForCurrentPathNonFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Document', folder2, id='new')
        self.uploadContentToPath(new, '', folder2)

    def testUploadCreateObjectOnSelfForCurrentPathFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Folder', folder2, id='new')
        self.uploadContentToPath(new, '', new)

    def testUploadCreateObjectOnRootForRootPathNonFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Document', folder2, id='new')
        self.uploadContentToPath(new, '/', self.portal)

    def testUploadCreateObjectOnRootForRootPathFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Folder', folder2, id='new')
        self.uploadContentToPath(new, '/', self.portal)

    def testUploadCreateObjectOnAbsoluteForAbsolutePathNonFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Document', self.portal, id='new')
        self.uploadContentToPath(new, '/folder1/folder2', folder2)

    def testUploadCreateObjectOnAbsoluteForAbsolutePathFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Folder', self.portal, id='new')
        self.uploadContentToPath(new, '/folder1/folder2', folder2)

    def testUploadCreateObjectOnRelativeForRelativePathNonFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Document', folder1, id='new')
        self.uploadContentToPath(new, 'folder2', folder2)

    def testUploadCreateObjectOnParentForInvalidRelativePathNonFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Document', folder1, id='new')
        self.uploadContentToPath(new, 'invalid', folder1)

    def testUploadCreateObjectOnSelfForInvalidRelativePathFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Folder', folder1, id='new')
        self.uploadContentToPath(new, 'invalid', new)

    def testUploadCreateObjectOnParentForInvalidAbsolutePathNonFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Document', folder1, id='new')
        self.uploadContentToPath(new, '/invalid', folder1)

    def testUploadCreateObjectOnSelfForInvalidAbsolutePathFolderish(self):
        folder1 = _createObjectByType('Folder', self.portal, id='folder1')
        folder2 = _createObjectByType('Folder', folder1, id='folder2')
        new = _createObjectByType('Folder', folder1, id='new')
        self.uploadContentToPath(new, '/invalid', new)


def test_suite():
    from unittest import TestSuite, makeSuite
    suite = TestSuite()
    suite.addTest(makeSuite(IntegrationTests))
    suite.addTest(ZopeTestCase.FunctionalDocFileSuite(
        'widget.txt',
        optionflags=URWTestCase.OPTIONFLAGS,
        package='archetypes.uploadreferencewidget.tests',
        test_class=URWTestCase.FunctionalTestCase))
    return suite


archetypes/uploadreferencewidget/demo.py
#
# archetypes.uploadreferencewidget
# Copyright(C), 2008, Enfold Systems, Inc. - ALL RIGHTS RESERVED
#
# This software is licensed under the Terms and Conditions
# contained within the "LICENSE.txt" file that accompanied
# this software.  Any inquiries concerning the scope or
# enforceability of the license should be addressed to:
#
# Enfold Systems, Inc.
# 4617 Montrose Blvd., Suite C215
# Houston, Texas 77006 USA
# p. +1 713.942.2377 | f. +1 832.201.8856
# www.enfoldsystems.com
# info@enfoldsystems.com
#

from Products.Archetypes.public import registerType
from Products.Archetypes.public import Schema
from Products.Archetypes.public import BaseSchema
from Products.Archetypes.public import BaseContent
from Products.Archetypes.public import ReferenceField
from archetypes.uploadreferencewidget.widget import UploadReferenceWidget
from archetypes.uploadreferencewidget import config

URWSchema = BaseSchema.copy() + Schema((

    ReferenceField(
        name='ref1',
        allowed_types=('File', 'Image'),
        relationship='rel1',
        required=True,
        widget=UploadReferenceWidget(
            label='Single Reference',
            description='This is the first field. It can reference a '
                        'single item. This field uses the site root as '
                        'the the startup directory.',
            startup_directory='/',
        ),
    ),

    ReferenceField(
        name='ref2',
        multiValued=True,
        allowed_types=('File', 'Image'),
        relationship='rel2',
        widget=UploadReferenceWidget(
            label='Multiple References',
            description='This is the second field. It can reference '
                        'multiple items. This field uses the current '
                        'folder as the startup directory.',
            startup_directory='',
        ),
    ),

    ReferenceField(
        name='ref3',
        allowed_types=('File', 'Image'),
        relationship='rel3',
        schemata='invalid',
        widget=UploadReferenceWidget(
            label='Invalid Path',
            description='This is the third field. It can reference a '
                        'single item. This field uses an invalid path '
                        'the the startup directory.',
            startup_directory='/non_existent',
        ),
    ),

))


class UploadReferenceWidgetDemo(BaseContent):
    """Demo content-type for testing UploadReferenceWidget."""
    schema = URWSchema


registerType(UploadReferenceWidgetDemo, config.PROJECTNAME)


archetypes/uploadreferencewidget/__init__.py
#
# archetypes.uploadreferencewidget
# Copyright(C), 2008, Enfold Systems, Inc. - ALL RIGHTS RESERVED
#
# This software is licensed under the Terms and Conditions
# contained within the "LICENSE.txt" file that accompanied
# this software.  Any inquiries concerning the scope or
# enforceability of the license should be addressed to:
#
# Enfold Systems, Inc.
# 4617 Montrose Blvd., Suite C215
# Houston, Texas 77006 USA
# p. +1 713.942.2377 | f. +1 832.201.8856
# www.enfoldsystems.com
# info@enfoldsystems.com
#

from Products.CMFCore import utils
from Products.CMFCore.permissions import AddPortalContent
from Products.CMFCore.DirectoryView import registerDirectory

from Products.Archetypes.public import listTypes
from Products.Archetypes.public import process_types

from archetypes.uploadreferencewidget import config
from archetypes.uploadreferencewidget import demo

# This patch can be removed when the Archetypes 1.6/1.5.5/1.4.6
# versions are released
import patch


def initialize(context):

    content_types, constructors, ftis = process_types(
        listTypes(config.PROJECTNAME), config.PROJECTNAME)

    utils.ContentInit(
        config.PROJECTNAME + ' Content',
        content_types=content_types,
        permission=AddPortalContent,
        extra_constructors=constructors,
        fti=ftis,
    ).initialize(context)


archetypes/uploadreferencewidget/skins/uploadreferencewidget/uploadreference.js
/*
* Utilities for the Upload Reference Widget
*/

function addUploadReferenceInput(field) {
    var place = document.getElementById(field + '_inputs');
    var br = document.createElement('br');
    place.appendChild(br);
    var input = document.createElement('input');
    input.size = '30';
    input.type = 'file';
    input.name = field + '_file:list';
    place.appendChild(input);
    return false;
}

function showUploadReference(field) {
    document.getElementById('box_' + field + '_select').style.display = 'none';
    document.getElementById('box_' + field + '_upload').style.display = 'block';
}

function hideUploadReference(field) {
    document.getElementById('box_' + field + '_upload').style.display = 'none';
    document.getElementById('box_' + field + '_select').style.display = 'block';
}


archetypes/uploadreferencewidget/widget.py
#
# archetypes.uploadreferencewidget
# Copyright(C), 2008, Enfold Systems, Inc. - ALL RIGHTS RESERVED
#
# This software is licensed under the Terms and Conditions
# contained within the "LICENSE.txt" file that accompanied
# this software.  Any inquiries concerning the scope or
# enforceability of the license should be addressed to:
#
# Enfold Systems, Inc.
# 4617 Montrose Blvd., Suite C215
# Houston, Texas 77006 USA
# p. +1 713.942.2377 | f. +1 832.201.8856
# www.enfoldsystems.com
# info@enfoldsystems.com
#

import mimetypes
from Acquisition import aq_parent
from OFS.interfaces import IFolder
from AccessControl import ClassSecurityInfo
from zope.component import queryUtility
from Products.CMFCore.utils import getToolByName
from Products.Archetypes.Registry import registerWidget
from archetypes.referencebrowserwidget.widget import ReferenceBrowserWidget

import pkg_resources
try:
    from z3c.blobfile.interfaces import IBlobFile, IBlobImage
    HAVE_BLOBS = True
except ImportError:
    HAVE_BLOBS = False

try:
    pkg_resources.get_distribution('plone.dexterity')
except pkg_resources.DistributionNotFound:
    HAS_DEXTERITY = False
    pass
else:
    HAS_DEXTERITY = True
    from plone.dexterity.interfaces import IDexterityFTI
    from plone.dexterity.utils import createContent, addContentToContainer

    if HAVE_BLOBS:
        from plone.namedfile import NamedBlobImage as Image
        from plone.namedfile import NamedBlobFile as File
    else:
        from plone.namedfile import NamedImage as Image
        from plone.namedfile import NamedFile as File


class UploadReferenceWidget(ReferenceBrowserWidget):

    _properties = ReferenceBrowserWidget._properties.copy()
    _properties.update({
        'macro': 'uploadreference',
        'helper_js': (
            'uploadreference.js',
            'referencebrowser.js',
            'select_lists.js',
        ),
    })

    security = ClassSecurityInfo()

    security.declarePublic('process_form')
    def process_form(self, instance, field, form, empty_marker=None,
                     emptyReturnsMarker=False, validating=True):
        """A custom implementation for the widget form processing."""
        if validating:
            # At the validation phase, only return the existing UID field
            # value or the filename of the new FileUpload object
            value = form.get(field.getName())
            if value:
                return value, {}

            value = empty_marker
            files_list = form.get('%s_file' % field.getName(), [])
            if files_list:
                obj = files_list[0]
                value = getattr(obj, 'filename', getattr(obj, 'name', ''))
            return value, {}

        option = form.get('%s_option' % field.getName(), empty_marker)
        value = form.get(field.getName(), empty_marker)

        # In this case, just do the trivial
        if option == 'select':
            if value is empty_marker:
                return empty_marker
            if emptyReturnsMarker and value == '':
                return empty_marker
            return value, {}

        # In this case, we need to create new object(s) and get its UID(s)
        if option == 'upload':
            result = []
            files_list = form.get('%s_file' % field.getName(), [])
            portal = getToolByName(instance, 'portal_url').getPortalObject()

            # Define the destination folder
            root = instance
            foldername = self.getStartupDirectory(instance, field)
            if foldername.startswith('/'):
                root = portal
                foldername = foldername[1:]
            try:
                folder = root.restrictedTraverse(foldername)
            except (KeyError, AttributeError):
                # If the startup_directory doesn't exists, fallback to
                # the current instance if folderish, or its parent
                if IFolder.providedBy(instance):
                    folder = instance
                else:
                    folder = aq_parent(instance)

            for fileobj in files_list:
                # Normalize the filename
                filename = getattr(fileobj, 'filename', '')
                filename = filename.split('\\')[-1]

                if filename:
                    # Guess the mimetype & define the content-type class
                    content = 'File'
                    mimetype = mimetypes.guess_type(filename)[0] or ""
                    if mimetype.startswith('image'):
                        content = 'Image'

                    ttol = getToolByName(instance, "portal_types")
                    fti = ttol.get(content)
                    # Create the new content
                    if HAS_DEXTERITY and IDexterityFTI.providedBy(fti):
                        filename = filename.decode('utf-8')
                        obj = createContent(content, id=filename, title=filename)
                        obj = addContentToContainer(folder, obj)
                        fileobj.seek(0)
                        data = fileobj.read()
                        headers = fileobj.headers
                        contentType = 'application/octet-stream'
                        if headers:
                            contentType = headers.get('Content-Type', contentType)
                        obj.filename = filename
                        if content == 'File':
                            obj.file = File(data, contentType=contentType, filename=filename)
                        else:
                            obj.image = Image(data, contentType=contentType, filename=filename)
                    else:
                        # Create Content with Archtypes
                        old_id = folder.generateUniqueId(content)
                        new_id = folder.invokeFactory(content, id=old_id, title=filename)
                        obj = getattr(folder, new_id)
                        obj._renameAfterCreation()
                        obj.unmarkCreationFlag()
                        obj.update_data(fileobj, mimetype)
                    obj.reindexObject()

                    result.append(obj.UID())

            if field.multiValued:
                # Multi valued, append the old value
                result.extend(value)
            elif result:
                # Non multi valued, with a valid upload, return it
                result = result[0]
            else:
                # Non multi valued, with an invalid upload, return the old value
                result = value

            return result, {}

        return empty_marker


registerWidget(
    UploadReferenceWidget,
    title='Upload Reference Widget',
    description='A widget that allows you to upload a content '
                'and create a reference to it.',
    used_for=('Products.Archetypes.Field.ReferenceField',)
)
